solidity/random-beacon/test/ReimbursmentPool.test.ts

```diff
@@ -0,0 +1,517 @@
+          thirdParty.address
+        )
+        const thirdPartyBalanceDiff = thirdPartyBalanceAfter.sub(
+          thirdPartyBalanceBefore
+        )
+        expect(thirdPartyBalanceDiff).to.be.equal(
+          ethers.utils.parseEther("2.0")
+        )
+      })
+
+      it("should emit FundsWithdrawn event", async () => {
+        await expect(
+          reimbursementPool
+            .connect(owner)
+            .withdraw(ethers.utils.parseEther("2.0"), thirdParty.address)
+        )
+          .to.emit(reimbursementPool, "FundsWithdrawn")
+          .withArgs(ethers.utils.parseEther("2.0"), thirdParty.address)
+      })
+    })
+
+    context("when receiver is zero address", () => {
+      it("should revert", async () => {
+        await expect(
+          reimbursementPool.connect(owner).withdraw(42, ZERO_ADDRESS)
+        ).to.be.revertedWith("Receiver's address cannot be zero")
+      })
+    })
+
+    context("when withdrawing more than the pool's balance", () => {
+      it("should revert", async () => {
+        await expect(
+          reimbursementPool
+            .connect(owner)
+            .withdraw(ethers.utils.parseEther("42.0"), ZERO_ADDRESS)
+        ).to.be.revertedWith("Insufficient contract balance")
+      })
+    })
+  })
+
+  describe("refund", () => {
+    beforeEach(async () => {
+      await createSnapshot()
+
+      await thirdParty.sendTransaction({
+        to: reimbursementPool.address,
+        value: ethers.utils.parseEther("10.0"), // Send 10.0 ETH
+      })
+    })
+
+    afterEach(async () => {
+      await restoreSnapshot()
+    })
+
+    context("when contract is not authorized", () => {
+      it("should revert", async () => {
+        await expect(
+          reimbursementPool
+            .connect(thirdParty)
+            .refund(ethers.utils.parseEther("2.0"), thirdParty.address)
+        ).to.be.revertedWith("Contract is not authorized for a refund")
+      })
+    })
+
+    context("when contract is authorized", () => {
+      beforeEach(async () => {
+        await createSnapshot()
+
+        await reimbursementPool
+          .connect(owner)
+          .authorize(thirdPartyContract.address)
+      })
+
+      afterEach(async () => {
+        await restoreSnapshot()
+      })
+
+      context("when tx gas price is lower than the max gas price", () => {
+        it("should refund based on tx.gasprice", async () => {
+          const refundeeBalanceBefore = await provider.getBalance(
+            refundee.address
+          )
+
+          await reimbursementPool
+            .connect(thirdPartyContract)
+            .refund(50000, refundee.address)
+
+          const refundeeBalanceAfter = await provider.getBalance(
+            refundee.address
+          )
+          const refundeeBalanceDiff = refundeeBalanceAfter.sub(
+            refundeeBalanceBefore
+          )
+          expect(refundeeBalanceDiff).to.be.gt(
+            ethers.utils.parseUnits("73000", "gwei")
```

Comment:
I've added more descriptions. The idea of the lower range at the moment of writing tests was to check if the gas is greater than 73000 assuming the gas price was 1gwei. It was just to show it's not 0, but I see that it might be a bit confusing, so I changed it just to `>0`. The higher range is calculated based on: `gas*tx.gasprice` which is a bit less than 145k. The precise number in the other test is based on the `maxGasPrice` which is set to 1gwei for testing purposes.